### PR TITLE
Add set committed QC function in forensics

### PR DESCRIPTION
**Pull request by wjrjerome**

### CODE REVIEW

File: consensus/tests/engine_v2_tests/forensics_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	var decodedExtraField utils.ExtraFields_v2
+	// Decode the qc1 and qc2
+	err := utils.DecodeBytesExtraFields(currentBlock.Header().Extra, &decodedExtraField)
+	assert.Nil(t, err)
+	err = forensics.SetCommittedQCs(append(headers, *blockchain.GetHeaderByNumber(903), *blockchain.GetHeaderByNumber(902)), *decodedExtraField.QuorumCert)
+	assert.NotNil(t, err)
+	assert.Equal(t, "Headers shall be on the same chain and in the right order", err.Error())
+
+	err = forensics.SetCommittedQCs(append(headers, *blockchain.GetHeaderByNumber(903), *blockchain.GetHeaderByNumber(904)), *decodedExtraField.QuorumCert)
+	assert.Nil(t, err)
```
Comment (wgr523) [inline consensus/tests/engine_v2_tests/forensics_test.go:97]: I think at last, add an assert for `HighestCommittedQCs` it better.

**Review by wgr523 — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 2]
```diff
@@ -930,6 +930,10 @@ func (x *XDPoS_v2) commitBlocks(blockChainReader consensus.ChainReader, proposed
 			Round:  round,
 		}
 		log.Debug("Successfully committed block", "Committed block Hash", x.highestCommitBlock.Hash, "Committed round", x.highestCommitBlock.Round)
+		// Perform forensics related operation
+		var headerQcToBeCommitted []types.Header
+		headerQcToBeCommitted = append(headerQcToBeCommitted, *parentBlock, *proposedBlockHeader)
+		go x.forensics.SetCommittedQCs(headerQcToBeCommitted, *incomingQc)
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go:936]: don't understand why you use array to send the parentBlock and proposedBlockheader, the order matters in your implementation, and just 2 parameters. Unless you are preparing for future multiple headers case.

File: consensus/XDPoS/engines/engine_v2/forensics.go [hunk 3]
```diff
@@ -17,36 +21,68 @@ type ForensicProof struct {
...
+// Set the forensics committed QCs list. The order is from grandparent to current header. i.e it shall follow the QC in its header as follow [hcqc1, hcqc2, hcqc3]
+func (f *Forensics) SetCommittedQCs(headers []types.Header, incomingQC utils.QuorumCert) error {
+	// highestCommitQCs is an array, assign the parentBlockQc and its child as well as its grandchild QC into this array for forensics purposes.
+	if len(headers) != 2 {
+		log.Error("[SetCommittedQcs] Received input length not equal to 2", len(headers))
+		return fmt.Errorf("Received headers length not equal to 2 ")
+	}
+
+	var committedQCs []utils.QuorumCert
+	for i, h := range headers {
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/forensics.go:57]: you can use below loop to remove `i != 0` 
```
for i := 1; i < 2; i++ {
```

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 2 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go:936]: Correct. Array is more flexible than passing parameters. I also have verification in place to check the order is correct in the array.

File: consensus/XDPoS/engines/engine_v2/forensics.go (hunk 3 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/forensics.go:57]: the i != 0 is just one condition. I still need to loop through from index 0

**Pull request merged**